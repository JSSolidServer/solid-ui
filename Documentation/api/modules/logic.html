<!doctype html>
<html class="default no-js">
<head>
	<meta charset="utf-8">
	<meta http-equiv="X-UA-Compatible" content="IE=edge">
	<title>logic | solid-ui</title>
	<meta name="description" content="Documentation for solid-ui">
	<meta name="viewport" content="width=device-width, initial-scale=1">
	<link rel="stylesheet" href="../assets/css/main.css">
	<script async src="../assets/js/search.js" id="search-script"></script>
</head>
<body>
<header>
	<div class="tsd-page-toolbar">
		<div class="container">
			<div class="table-wrap">
				<div class="table-cell" id="tsd-search" data-index="../assets/js/search.json" data-base="..">
					<div class="field">
						<label for="tsd-search-field" class="tsd-widget search no-caption">Search</label>
						<input id="tsd-search-field" type="text" />
					</div>
					<ul class="results">
						<li class="state loading">Preparing search index...</li>
						<li class="state failure">The search index is not available</li>
					</ul>
					<a href="../index.html" class="title">solid-ui</a>
				</div>
				<div class="table-cell" id="tsd-widgets">
					<div id="tsd-filter">
						<a href="#" class="tsd-widget options no-caption" data-toggle="options">Options</a>
						<div class="tsd-filter-group">
							<div class="tsd-select" id="tsd-filter-visibility">
								<span class="tsd-select-label">All</span>
								<ul class="tsd-select-list">
									<li data-value="public">Public</li>
									<li data-value="protected">Public/Protected</li>
									<li data-value="private" class="selected">All</li>
								</ul>
							</div>
							<input type="checkbox" id="tsd-filter-inherited" checked />
							<label class="tsd-widget" for="tsd-filter-inherited">Inherited</label>
							<input type="checkbox" id="tsd-filter-externals" checked />
							<label class="tsd-widget" for="tsd-filter-externals">Externals</label>
						</div>
					</div>
					<a href="#" class="tsd-widget menu no-caption" data-toggle="menu">Menu</a>
				</div>
			</div>
		</div>
	</div>
	<div class="tsd-page-title">
		<div class="container">
			<ul class="tsd-breadcrumb">
				<li>
					<a href="../modules.html">solid-ui</a>
				</li>
				<li>
					<a href="logic.html">logic</a>
				</li>
			</ul>
			<h1>Module logic</h1>
		</div>
	</div>
</header>
<div class="container container-main">
	<div class="row">
		<div class="col-8 col-content">
			<section class="tsd-panel-group tsd-index-group">
				<h2>Index</h2>
				<section class="tsd-panel tsd-index-panel">
					<div class="tsd-index-content">
						<section class="tsd-index-section ">
							<h3>Variables</h3>
							<ul class="tsd-index-list">
								<li class="tsd-kind-variable tsd-parent-kind-module"><a href="logic.html#authn" class="tsd-kind-icon">authn</a></li>
								<li class="tsd-kind-variable tsd-parent-kind-module"><a href="logic.html#chat" class="tsd-kind-icon">chat</a></li>
								<li class="tsd-kind-variable tsd-parent-kind-module"><a href="logic.html#kb" class="tsd-kind-icon">kb</a></li>
								<li class="tsd-kind-variable tsd-parent-kind-module"><a href="logic.html#profile" class="tsd-kind-icon">profile</a></li>
								<li class="tsd-kind-variable tsd-parent-kind-module"><a href="logic.html#solidlogicsingleton" class="tsd-kind-icon">solid<wbr>Logic<wbr>Singleton</a></li>
								<li class="tsd-kind-variable tsd-parent-kind-module"><a href="logic.html#store" class="tsd-kind-icon">store</a></li>
							</ul>
						</section>
					</div>
				</section>
			</section>
			<section class="tsd-panel-group tsd-member-group ">
				<h2>Variables</h2>
				<section class="tsd-panel tsd-member tsd-kind-variable tsd-parent-kind-module">
					<a name="authn" class="tsd-anchor"></a>
					<h3><span class="tsd-flag ts-flagConst">Const</span> authn</h3>
					<div class="tsd-signature tsd-kind-icon">authn<span class="tsd-signature-symbol">:</span> <span class="tsd-signature-type">AuthnLogic</span><span class="tsd-signature-symbol"> = ...</span></div>
					<aside class="tsd-sources">
						<ul>
							<li>Defined in <a href="https://github.com/solid/solid-ui/blob/89676eb/src/logic.ts#L14">src/logic.ts:14</a></li>
						</ul>
					</aside>
				</section>
				<section class="tsd-panel tsd-member tsd-kind-variable tsd-parent-kind-module">
					<a name="chat" class="tsd-anchor"></a>
					<h3><span class="tsd-flag ts-flagConst">Const</span> chat</h3>
					<div class="tsd-signature tsd-kind-icon">chat<span class="tsd-signature-symbol">:</span> <span class="tsd-signature-type">ChatLogic</span><span class="tsd-signature-symbol"> = ...</span></div>
					<aside class="tsd-sources">
						<ul>
							<li>Defined in <a href="https://github.com/solid/solid-ui/blob/89676eb/src/logic.ts#L15">src/logic.ts:15</a></li>
						</ul>
					</aside>
				</section>
				<section class="tsd-panel tsd-member tsd-kind-variable tsd-parent-kind-module">
					<a name="kb" class="tsd-anchor"></a>
					<h3><span class="tsd-flag ts-flagConst">Const</span> kb</h3>
					<div class="tsd-signature tsd-kind-icon">kb<span class="tsd-signature-symbol">:</span> <span class="tsd-signature-type">LiveStore</span><span class="tsd-signature-symbol"> = ...</span></div>
					<aside class="tsd-sources">
						<ul>
							<li>Defined in <a href="https://github.com/solid/solid-ui/blob/89676eb/src/logic.ts#L12">src/logic.ts:12</a></li>
						</ul>
					</aside>
				</section>
				<section class="tsd-panel tsd-member tsd-kind-variable tsd-parent-kind-module">
					<a name="profile" class="tsd-anchor"></a>
					<h3><span class="tsd-flag ts-flagConst">Const</span> profile</h3>
					<div class="tsd-signature tsd-kind-icon">profile<span class="tsd-signature-symbol">:</span> <span class="tsd-signature-type">ProfileLogic</span><span class="tsd-signature-symbol"> = ...</span></div>
					<aside class="tsd-sources">
						<ul>
							<li>Defined in <a href="https://github.com/solid/solid-ui/blob/89676eb/src/logic.ts#L19">src/logic.ts:19</a></li>
						</ul>
					</aside>
				</section>
				<section class="tsd-panel tsd-member tsd-kind-variable tsd-parent-kind-module">
					<a name="solidlogicsingleton" class="tsd-anchor"></a>
					<h3><span class="tsd-flag ts-flagConst">Const</span> solid<wbr>Logic<wbr>Singleton</h3>
					<div class="tsd-signature tsd-kind-icon">solid<wbr>Logic<wbr>Singleton<span class="tsd-signature-symbol">:</span> <span class="tsd-signature-type">SolidLogic</span><span class="tsd-signature-symbol"> = ...</span></div>
					<aside class="tsd-sources">
						<ul>
							<li>Defined in <a href="https://github.com/solid/solid-ui/blob/89676eb/src/logic.ts#L8">src/logic.ts:8</a></li>
						</ul>
					</aside>
				</section>
				<section class="tsd-panel tsd-member tsd-kind-variable tsd-parent-kind-module">
					<a name="store" class="tsd-anchor"></a>
					<h3><span class="tsd-flag ts-flagConst">Const</span> store</h3>
					<div class="tsd-signature tsd-kind-icon">store<span class="tsd-signature-symbol">:</span> <span class="tsd-signature-type">LiveStore</span><span class="tsd-signature-symbol"> = ...</span></div>
					<aside class="tsd-sources">
						<ul>
							<li>Defined in <a href="https://github.com/solid/solid-ui/blob/89676eb/src/logic.ts#L11">src/logic.ts:11</a></li>
						</ul>
					</aside>
				</section>
			</section>
		</div>
		<div class="col-4 col-menu menu-sticky-wrap menu-highlight">
			<nav class="tsd-navigation primary">
				<ul>
					<li class=" ">
						<a href="../modules.html">Exports</a>
					</li>
					<li class="current tsd-kind-module">
						<a href="logic.html">logic</a>
					</li>
				</ul>
			</nav>
			<nav class="tsd-navigation secondary menu-sticky">
				<ul class="before-current">
					<li class=" tsd-kind-variable tsd-parent-kind-module">
						<a href="logic.html#authn" class="tsd-kind-icon">authn</a>
					</li>
					<li class=" tsd-kind-variable tsd-parent-kind-module">
						<a href="logic.html#chat" class="tsd-kind-icon">chat</a>
					</li>
					<li class=" tsd-kind-variable tsd-parent-kind-module">
						<a href="logic.html#kb" class="tsd-kind-icon">kb</a>
					</li>
					<li class=" tsd-kind-variable tsd-parent-kind-module">
						<a href="logic.html#profile" class="tsd-kind-icon">profile</a>
					</li>
					<li class=" tsd-kind-variable tsd-parent-kind-module">
						<a href="logic.html#solidlogicsingleton" class="tsd-kind-icon">solid<wbr>Logic<wbr>Singleton</a>
					</li>
					<li class=" tsd-kind-variable tsd-parent-kind-module">
						<a href="logic.html#store" class="tsd-kind-icon">store</a>
					</li>
				</ul>
			</nav>
		</div>
	</div>
</div>
<footer class="with-border-bottom">
	<div class="container">
		<h2>Legend</h2>
		<div class="tsd-legend-group">
			<ul class="tsd-legend">
				<li class="tsd-kind-variable"><span class="tsd-kind-icon">Variable</span></li>
				<li class="tsd-kind-function"><span class="tsd-kind-icon">Function</span></li>
				<li class="tsd-kind-function tsd-has-type-parameter"><span class="tsd-kind-icon">Function with type parameter</span></li>
				<li class="tsd-kind-type-alias"><span class="tsd-kind-icon">Type alias</span></li>
				<li class="tsd-kind-type-alias tsd-has-type-parameter"><span class="tsd-kind-icon">Type alias with type parameter</span></li>
			</ul>
			<ul class="tsd-legend">
				<li class="tsd-kind-interface"><span class="tsd-kind-icon">Interface</span></li>
			</ul>
			<ul class="tsd-legend">
				<li class="tsd-kind-class"><span class="tsd-kind-icon">Class</span></li>
			</ul>
		</div>
	</div>
</footer>
<div class="container tsd-generator">
	<p>Generated using <a href="https://typedoc.org/" target="_blank">TypeDoc</a></p>
</div>
<div class="overlay"></div>
<script src="../assets/js/main.js"></script>
</body>
</html>