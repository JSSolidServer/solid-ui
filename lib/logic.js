"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.store = exports.solidLogicSingleton = exports.profile = exports.kb = exports.chat = exports.authn = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var debug = _interopRequireWildcard(require("./debug"));

var _authSession = _interopRequireDefault(require("./authn/authSession"));

var _solidLogic = require("solid-logic");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

// This module of solid-ui has a main quadstore for the app to use
//
function thisFetch(_x, _x2) {
  return _thisFetch.apply(this, arguments);
}

function _thisFetch() {
  _thisFetch = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(url, requestInit) {
    var omitCreds;
    return _regenerator["default"].wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            omitCreds = requestInit && requestInit.credentials && requestInit.credentials === 'omit';

            if (!(_authSession["default"].info.webId && !omitCreds)) {
              _context.next = 5;
              break;
            }

            return _context.abrupt("return", _authSession["default"].fetch(url, requestInit));

          case 5:
            return _context.abrupt("return", window.fetch(url, requestInit));

          case 6:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  }));
  return _thisFetch.apply(this, arguments);
}

var solidLogicSingleton = new _solidLogic.SolidLogic({
  fetch: thisFetch
}, _authSession["default"]); // Make this directly accessible as it is what you need most of the time

exports.solidLogicSingleton = solidLogicSingleton;
var store = solidLogicSingleton.store;
exports.store = store;
var kb = store; // Very commonly used synonym of store - Knowledge Base

exports.kb = kb;
var authn = solidLogicSingleton.authn;
exports.authn = authn;
var chat = solidLogicSingleton.chat; // export const language = solidLogicSingleton.language // Does not work

exports.chat = chat;
var profile = solidLogicSingleton.profile;
exports.profile = profile;
debug.log('Unique quadstore initialized.'); // ends
//# sourceMappingURL=logic.js.map