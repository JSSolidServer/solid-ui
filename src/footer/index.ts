/*
    This file was copied from mashlib/src/global/footer.ts file. It is modified to
    work in solid-ui by adjusting where imported functions are found.
 */
import { IndexedFormula, NamedNode } from 'rdflib'
import { authSession, currentUser } from '../authn/authn'
import { addStyleClassToElement, getName, getPod, getPodOwner } from '../utils/headerFooterHelpers'

const DEFAULT_SOLID_PROJECT_URL = 'https://solidproject.org'
const DEFAULT_SOLID_PROJECT_NAME = 'solidproject.org'

/*
  FooterOptions allow for customizing the link and name of the link part of the footer.
  */
export type FooterOptions = {
    solidProjectUrl?: string,
    solidProjectName?: string
  }

/**
 * Initialize footer component, the footer object returned depends on whether the user is authenticated.
 * @param store the data store
 * @returns the footer
 */
export async function initFooter (store: IndexedFormula) {
  const footer = document.getElementById('PageFooter')
  if (!footer) {
    return
  }
  const pod = getPod()
  const podOwner = await getPodOwner(pod, store)
  rebuildFooter(footer, store, pod, podOwner)()
  authSession.onLogin(rebuildFooter(footer, store, pod, podOwner))
  authSession.onLogout(rebuildFooter(footer, store, pod, podOwner))
}
/**
 * @ignore exporting this only for the unit test
 */
export function rebuildFooter (footer: HTMLElement, store: IndexedFormula, pod: NamedNode | null, podOwner: NamedNode | null) {
  return async () => {
    const user = currentUser()
    footer.innerHTML = ''
    footer.appendChild(await createControllerInfoBlock(store, user, pod, podOwner))
  }
}
/**
 * @ignore exporting this only for the unit test
 */
export function createControllerInfoBlock (store: IndexedFormula, user: NamedNode | null, pod: NamedNode | null, podOwner: NamedNode | null): HTMLElement {
  const profileLinkContainer = document.createElement('div')
  if (!pod || !podOwner || (user && user.equals(podOwner))) {
    return profileLinkContainer
  }

  addStyleClassToElement(profileLinkContainer, ['footer-pod-info', 'footer'], 'footer')

  const podLinkPre = document.createElement('span')
  podLinkPre.innerText = "You're visiting "

  const podLink = document.createElement('a')
  podLink.href = pod.uri
  podLink.innerText = 'the Pod'

  const profileLinkPre = document.createElement('span')
  profileLinkPre.innerText = ' controlled by '

  const profileLink = document.createElement('a')
  profileLink.href = podOwner.uri
  profileLink.innerText = getName(store, podOwner)

  const solidProjectLinkPre = document.createElement('span')
  solidProjectLinkPre.innerText = '. For more info, check out '

  const solidProjectLink = document.createElement('a')
  solidProjectLink.href = DEFAULT_SOLID_PROJECT_URL
  solidProjectLink.innerText = DEFAULT_SOLID_PROJECT_NAME

  const solidProjectLinkPost = document.createElement('span')
  solidProjectLinkPost.innerText = '.'

  profileLinkContainer.appendChild(podLinkPre)
  profileLinkContainer.appendChild(podLink)
  profileLinkContainer.appendChild(profileLinkPre)
  profileLinkContainer.appendChild(profileLink)
  profileLinkContainer.appendChild(solidProjectLinkPre)
  profileLinkContainer.appendChild(solidProjectLink)
  profileLinkContainer.appendChild(solidProjectLinkPost)

  return profileLinkContainer
}
