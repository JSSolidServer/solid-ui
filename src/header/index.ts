/* global EventListenerOrEventListenerObject */
/*
    This file was copied from mashlib/src/global/header.ts file. It is modified to
    work in solid-ui by adjusting where imported functions are found.
 */
import { IndexedFormula, NamedNode } from 'rdflib'
import { icons } from '..'
import { loginStatusBox, authSession, currentUser } from '../authn/authn'
import * as widgets from '../widgets'
import { emptyProfile } from './empty-profile'
import { addStyleClassToElement, getPod, throttle } from './headerHelpers'

export type MenuItemLink = {
  label: string,
  url: string
}

export type MenuItemButton = {
  label: string,
  onclick: () => {}
}

export type MenuItems = MenuItemLink | MenuItemButton

/*
HeaderOptions allow for customizing the logo and menu list.  If a logo is not provided the default
    is solid. Menulist will always show a link to logout and to the users profile.
  */
export type HeaderOptions = {
  logo?: string,
  menuList?: MenuItems[]
}

/**
 * Initialize header component, the header object returned depends on whether the user is authenticated.
 * @param store the data store
 * @param options allow the header to be customized with a personalized logo and a menu list of links or buttons.
 * @returns a header for an authenticated user with menu items given or a login screen
 */
export async function initHeader (store: IndexedFormula, options?: HeaderOptions) {
  const header = document.getElementById('PageHeader')
  if (!header) {
    return
  }

  const pod = getPod()
  rebuildHeader(header, store, pod, options)()
  authSession.onLogout(rebuildHeader(header, store, pod, options))
  authSession.onLogin(rebuildHeader(header, store, pod, options))
}
/**
 * @ignore exporting this only for the unit test
 */
export function rebuildHeader (header: HTMLElement, store: IndexedFormula, pod: NamedNode, options?: HeaderOptions) {
  return async () => {
    const user = currentUser()
    header.innerHTML = ''
    header.appendChild(await createBanner(store, pod, user, options))
  }
}
/**
 * @ignore exporting this only for the unit test
 */
export async function createBanner (store: IndexedFormula, pod: NamedNode, user: NamedNode | null, options?: HeaderOptions): Promise<HTMLElement> {
  const podLink = document.createElement('a')
  podLink.href = pod.uri
  addStyleClassToElement(podLink, ['header-banner__link'])
  const image = document.createElement('img')
  if (options) {
    image.src = options.logo ? options.logo : 'https://solidproject.org/assets/img/solid-emblem.svg'
  }
  addStyleClassToElement(image, ['header-banner__icon'])
  podLink.appendChild(image)

  const userMenu = user
    ? await createUserMenu(store, user, options)
    : createLoginSignUpButtons()

  const helpMenu = createHelpMenu()

  const banner = document.createElement('div')
  addStyleClassToElement(banner, ['header-banner'])
  banner.appendChild(podLink)

  const leftSideOfHeader = document.createElement('div')
  addStyleClassToElement(leftSideOfHeader, ['header-banner__right-menu'])
  leftSideOfHeader.appendChild(userMenu)
  leftSideOfHeader.appendChild(helpMenu)

  banner.appendChild(leftSideOfHeader)

  return banner
}
/**
 * @ignore exporting this only for the unit test
 */
export function createHelpMenu () {
  const helpMenuList = document.createElement('ul')
  addStyleClassToElement(helpMenuList, ['header-user-menu__list'])
  helpMenuList.appendChild(createUserMenuItem(createUserMenuLink('User guide', 'https://github.com/solid/userguide')))
  helpMenuList.appendChild(createUserMenuItem(createUserMenuLink('Report a problem', 'https://github.com/solid/solidos/issues')))

  const helpMenu = document.createElement('nav')

  addStyleClassToElement(helpMenu, ['header-user-menu__navigation-menu'])
  helpMenu.setAttribute('aria-hidden', 'true')
  helpMenu.appendChild(helpMenuList)

  const helpIcon = icons.iconBase + 'noun_144.svg'

  const helpMenuContainer = document.createElement('div')
  addStyleClassToElement(helpMenuContainer, ['header-banner__user-menu'])
  addStyleClassToElement(helpMenuContainer, ['header-user-menu'])
  helpMenuContainer.appendChild(helpMenu)

  const helpMenuTrigger = document.createElement('button')
  addStyleClassToElement(helpMenuTrigger, ['header-user-menu__trigger'])
  helpMenuTrigger.type = 'button'

  const helpMenuIcon = document.createElement('img')
  helpMenuIcon.src = helpIcon
  addStyleClassToElement(helpMenuIcon, ['header-banner__help-icon'])
  helpMenuContainer.appendChild(helpMenuTrigger)
  helpMenuTrigger.appendChild(helpMenuIcon)

  const throttledMenuToggle = throttle((event: Event) => toggleMenu(event, helpMenuTrigger, helpMenu), 50)
  helpMenuTrigger.addEventListener('click', throttledMenuToggle)
  let timer = setTimeout(() => null, 0)
  helpMenuContainer.addEventListener('mouseover', event => {
    clearTimeout(timer)
    throttledMenuToggle(event)
  })
  helpMenuContainer.addEventListener('mouseout', event => {
    timer = setTimeout(() => throttledMenuToggle(event), 200)
  })

  return helpMenuContainer
}
/**
 * @ignore exporting this only for the unit test
 */
export function createLoginSignUpButtons () {
  const profileLoginButtonPre = document.createElement('div')
  addStyleClassToElement(profileLoginButtonPre, ['header-banner__login'])
  profileLoginButtonPre.appendChild(loginStatusBox(document, null, {}))
  return profileLoginButtonPre
}
/**
 * @ignore exporting this only for the unit test
 */
export function createUserMenuButton (label: string, onClick: EventListenerOrEventListenerObject): HTMLElement {
  const button = document.createElement('button')
  addStyleClassToElement(button, ['header-user-menu__button'])
  button.addEventListener('click', onClick)
  button.innerText = label
  return button
}
/**
 * @ignore exporting this only for the unit test
 */
export function createUserMenuLink (label: string, href: string): HTMLElement {
  const link = document.createElement('a')
  addStyleClassToElement(link, ['header-user-menu__link'])
  link.href = href
  link.innerText = label
  return link
}

/**
 * @ignore exporting this only for the unit test
 */
export async function createUserMenu (store: IndexedFormula, user: NamedNode, options?: HeaderOptions): Promise<HTMLElement> {
  const fetcher = (<any>store).fetcher
  if (fetcher) {
    // Making sure that Profile is loaded before building menu
    await fetcher.load(user)
  }
  const loggedInMenuList = document.createElement('ul')
  addStyleClassToElement(loggedInMenuList, ['header-user-menu__list'])
  loggedInMenuList.appendChild(createUserMenuItem(createUserMenuLink('Show your profile', user.uri)))
  if (options) {
    if (options.menuList) {
      options.menuList.forEach(function (menuItem) {
        const menuItemType: string = (menuItem as MenuItemLink).url ? 'url' : 'onclick'
        if (menuItemType === 'url') {
          loggedInMenuList.appendChild(createUserMenuItem(createUserMenuLink(menuItem.label, menuItem[menuItemType])))
        } else {
          loggedInMenuList.appendChild(createUserMenuItem(createUserMenuButton(menuItem.label, menuItem[menuItemType])))
        }
      })
    }
  }

  loggedInMenuList.appendChild(createUserMenuItem(createUserMenuButton('Log out', () => authSession.logout())))

  const loggedInMenu = document.createElement('nav')

  addStyleClassToElement(loggedInMenu, ['header-user-menu__navigation-menu'])
  loggedInMenu.setAttribute('aria-hidden', 'true')
  loggedInMenu.appendChild(loggedInMenuList)

  const loggedInMenuTrigger = document.createElement('button')
  addStyleClassToElement(loggedInMenuTrigger, ['header-user-menu__trigger'])
  loggedInMenuTrigger.type = 'button'
  const profileImg = getProfileImg(store, user)
  if (typeof profileImg === 'string') {
    loggedInMenuTrigger.innerHTML = profileImg
  } else {
    loggedInMenuTrigger.appendChild(profileImg)
  }

  const loggedInMenuContainer = document.createElement('div')
  addStyleClassToElement(loggedInMenuContainer, ['header-banner__user-menu'])
  addStyleClassToElement(loggedInMenuContainer, ['header-user-menu'])
  loggedInMenuContainer.appendChild(loggedInMenuTrigger)
  loggedInMenuContainer.appendChild(loggedInMenu)

  const throttledMenuToggle = throttle((event: Event) => toggleMenu(event, loggedInMenuTrigger, loggedInMenu), 50)
  loggedInMenuTrigger.addEventListener('click', throttledMenuToggle)
  let timer = setTimeout(() => null, 0)
  loggedInMenuContainer.addEventListener('mouseover', event => {
    clearTimeout(timer)
    throttledMenuToggle(event)
  })
  loggedInMenuContainer.addEventListener('mouseout', event => {
    timer = setTimeout(() => throttledMenuToggle(event), 200)
  })

  return loggedInMenuContainer
}

/**
 * @ignore exporting this only for the unit test
 */
export function createUserMenuItem (child: HTMLElement): HTMLElement {
  const menuProfileItem = document.createElement('li')
  addStyleClassToElement(menuProfileItem, ['header-user-menu__list-item'])
  menuProfileItem.appendChild(child)
  return menuProfileItem
}
/**
 * @ignore exporting this only for the unit test
 */
export function getProfileImg (store: IndexedFormula, user: NamedNode): string | HTMLElement {
  const profileUrl = null
  try {
    const profileUrl = widgets.findImage(user)
    if (!profileUrl) {
      return emptyProfile
    }
  } catch {
    return emptyProfile
  }

  const profileImage = document.createElement('div')
  addStyleClassToElement(profileImage, ['header-user-menu__photo'])
  profileImage.style.backgroundImage = `url("${profileUrl}")`
  return profileImage
}

/**
 * @internal
 */
function toggleMenu (event: Event, trigger: HTMLButtonElement, menu: HTMLElement): void {
  const isExpanded = trigger.getAttribute('aria-expanded') === 'true'
  const expand = event.type === 'mouseover'
  const close = event.type === 'mouseout'
  if ((isExpanded && expand) || (!isExpanded && close)) {
    return
  }
  trigger.setAttribute('aria-expanded', (!isExpanded).toString())
  menu.setAttribute('aria-hidden', isExpanded.toString())
}
