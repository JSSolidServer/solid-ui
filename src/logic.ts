// This module of solid-ui has a main quadstore for the app to use
//

import * as debug from './debug'
import authSession from './authn/authSession'
import { SolidLogic } from 'solid-logic'

async function thisFetch (url, requestInit) {
  const omitCreds = requestInit && requestInit.credentials && requestInit.credentials == 'omit'
  if (authSession.info.webId && !omitCreds) { // see https://github.com/solid/solidos/issues/114
    // In fact ftech should respect crentials omit itself
    return authSession.fetch(url, requestInit)
  } else {
    return window.fetch(url, requestInit)
  }
}

export const solidLogicSingleton = new SolidLogic({ fetch: thisFetch }, authSession)

// Make this directly accessible as it is what you need most of the time
export const store = solidLogicSingleton.store
export const kb = store // Very commonly used synonym of store - Knowledge Base

export const authn = solidLogicSingleton.authn
export const chat = solidLogicSingleton.chat

// export const language = solidLogicSingleton.language // Does not work

export const profile = solidLogicSingleton.profile

debug.log('Unique quadstore initialized.')

// ends
