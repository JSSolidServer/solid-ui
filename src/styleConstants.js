export default {
  highlightColor: '#7C4DFF', // Solid lavender https://design.inrupt.com/atomic-core/?cat=Core

  formBorderColor: '#888888', // Mid-grey
  formHeadingColor: '#888888', // originally was brown; now grey
  lowProfileLinkColor: '#3B5998', // Grey-blue, e.g., for field labels linking to ontology
  formFieldNameBoxWidth: '8em', // The fixed amount to get form fields to line up
  // The latter we put in when switching awy from using tables.  Getting alignment between
  // fields in different groups though is hard problem.

  mediaModuleCanvasWidth: '640',
  mediaModuleCanvasHeight: '480',

  textInputSize: 20, // Default text input size in characters roughly
  tabBorderRadius: '0.2em',

  textInputBackgroundColor: '#eef',
  textInputBackgroundColorUneditable: '#fff',
  textInputColor: '#000',
  textInputColorPending: '#888',

  defaultErrorBackgroundColor: '#fee'
}
