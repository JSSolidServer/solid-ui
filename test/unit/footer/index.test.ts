import { silenceDebugMessages } from '../../helpers/setup'
import {
  initFooter,
  rebuildFooter,
  createControllerInfoBlock
} from '../../../src/footer'
import { NamedNode } from 'rdflib'
// @ts-ignore
import { solidLogicSingleton } from '../../../src/logic'

const store = solidLogicSingleton.store

silenceDebugMessages()

describe('footer', () => {
  it('is exposed on public API', async () => {
    expect(await initFooter(store)).toMatchSnapshot()
  })
})
describe('rebuildFooter', () => {
  it('creates a link', () => {
    const footer = document.createElement('div')
    const pod = new NamedNode('https://test.com')
    const podOwner = new NamedNode('https://test2.com')
    expect(rebuildFooter(footer, store, pod, podOwner)).toMatchSnapshot()
  })
  it('does NOT creates a link in the footer', () => {
    const footer = document.createElement('div')
    const pod = new NamedNode('https://test.com')
    const podOwner = new NamedNode('https://test.com')
    expect(rebuildFooter(footer, store, pod, podOwner)).toMatchSnapshot()
  })
})
describe('createControllerInfoBlock', () => {
  it('runs createControllerInfoBlock', async () => {
    const pod = new NamedNode('https://test.com')
    const podOwner = new NamedNode('https://test.com')
    const user = new NamedNode('https://sharonstrats.inrupt.net/profile/card#me')
    expect(createControllerInfoBlock(store, user, pod, podOwner)).toMatchSnapshot()
  })
  it('assigns class to footer', async () => {
    const pod = new NamedNode('https://test.com')
    const podOwner = new NamedNode('https://test.com')
    const user = new NamedNode('https://sharonstrats.inrupt.net/profile/card#me')
    const footer = createControllerInfoBlock(store, user, pod, podOwner)
    expect(footer.className).toContain('footer-pod-info')
    expect(footer.className).toContain('footer')
  })
})
