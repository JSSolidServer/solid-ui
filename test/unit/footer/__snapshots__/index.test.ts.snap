// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createControllerInfoBlock makes use of options 1`] = `
<div
  class="footer-pod-info-0-1-5 footer-0-1-6"
>
  <span />
  <a
    href="https://test.com"
  />
  <span />
  <a
    href="https://test.com"
  />
  <span />
  <a
    href="https://solid.com/"
  />
  <span />
</div>
`;

exports[`createControllerInfoBlock runs createControllerInfoBlock 1`] = `
<div
  class="footer-pod-info-0-1-1 footer-0-1-2"
>
  <span />
  <a
    href="https://test.com"
  />
  <span />
  <a
    href="https://test.com"
  />
  <span />
  <a
    href="https://solidproject.org"
  />
  <span />
</div>
`;

exports[`footer is exposed on public API 1`] = `undefined`;

exports[`rebuildFooter creates a link 1`] = `[Function]`;

exports[`rebuildFooter does NOT creates a link in the footer 1`] = `[Function]`;
